cpp/ql/src/experimental/Security/CWE/CWE-020/NoCheckBeforeUnsafePutUser.ql

```diff
@@ -16,44 +16,143 @@
+/*
+ * using functions' name to model linux syscall which usually
+ * starts with prefix like:
+ *   C_SYSC_
+ *   SYSC_
+ *   compat_Sys_
+ *   Sys_
+ */
+
+class SysCallFunction extends Function {
+  SysCallFunction() {
+    this.getName().regexpMatch("C_SYSC_[a-zA-Z]+") or
+    this.getName().regexpMatch("SYSC_[a-zA-Z]+") or
+    this.getName().regexpMatch("compat_SyS_[a-zA-Z]+") or
+    this.getName().regexpMatch("SyS_[a-zA-Z]+")
+  }
+}
+
+class SysCallParameter extends Parameter {
+  SysCallParameter() { exists(SysCallFunction f | f.getAParameter() = this) }
+}
+
+class UserModePtrCheckMacro extends Macro {
   VariableAccess va;
 
-  WriteAccessCheckMacro() {
+  UserModePtrCheckMacro() {
     this.getName() = ["user_write_access_begin", "user_access_begin", "access_ok"] and
     va.getEnclosingElement() = this.getAnInvocation().getAnExpandedElement()
   }
 
   VariableAccess getArgument() { result = va }
 }
 
+/*
+ *   It's really difficult to spot the second argument passed to
+ *   unsafe_put_user and used as user-mode pointer. In previous
+ *   version PointerDereferenceExpr is used to model user-mode
+ *   pointer, but in some odd situation this may lead to false
+ *   positive. For example:
+ *
+ *   void function foo(struct some_struct* __user pointer, long* dummy)
+ *   {
+ *       if(!access_ok(pointer))            //security check for pointer
+ *           return;
+ *       unsafe_put_user(*dummy, pointer);
+ *   }
+ *
+ *   parameter dummy will be regared as user-mode pointer used
+ *   in unsafe_put_user without security check using access_ok
+ *   but in fact dummy is only used to read memory otherwise
+ *   instead of wring user-mode memory.
+ *
+ *   Reading implementation of unsafe_put_user in Linux codebase,
+ *   I found that there is a sizeof operation for user-mode pointer.
+ *   So SizeofExprOperator is used instead to model user-mode pointer.
+ */
+
 class UnSafePutUserMacro extends Macro {
-  PointerDereferenceExpr writeUserPtr;
+  SizeofExprOperator sizeOfOperator;
 
   UnSafePutUserMacro() {
     this.getName() = "unsafe_put_user" and
-    writeUserPtr.getEnclosingElement() = this.getAnInvocation().getAnExpandedElement()
+    sizeOfOperator.getEnclosingElement() = this.getAnInvocation().getAnExpandedElement()
+  }
+
+  Expr getExprOperand() {
+    result =
+      sizeOfOperator
+          .getExprOperand()
+          .getAChild()
+          .(AddressOfExpr)
+          .getOperand()
+          .(FieldAccess)
+          .getQualifier() or
+    result = sizeOfOperator.getExprOperand().getAChild()
+  }
+}
+
+/*
+ *  Since there is no convenient way to indentify user mode pointer,
+ *  In this query syscall parameter or function return value will
+ *  be regared source node that's used as user mode pointer
+ */
+
+class UserModePtrNode extends DataFlow::Node {
+  UserModePtrNode() {
+    exists(SysCallParameter p | this.asParameter() = p) or
+    exists(FunctionCall va | this.asExpr() = va)
```

Comment:
Should this line match only a call to a `SysCallFunction`?

Reply:
> Should this line match only a call to a `SysCallFunction`?

In this query I model taint souce comes from Linux syscall and other function return values that maybe used as user-mode pointer. So this line SysCallParameter used to model parameter from linux syscall parameter, and FunctionCall used to model other function return values.

Reply:
This is still matching every function call in the program.
```suggestion
    exists(FunctionCall va | this.asExpr() = va and va.getTarget() instanceof SysCallFunction)
```